src/query_sink.rs

```diff
@@ -0,0 +1,129 @@
+use winapi::{
+    um::wbemcli::{
+        IWbemClassObject,
+        WBEM_NO_ERROR,
+    },
+    shared::{
+        ntdef::HRESULT,
+        wtypes::BSTR,
+    },
+    ctypes::{
+        c_long,
+    },
+};
+use winapi::um::wbemcli::{IWbemObjectSink, IWbemObjectSinkVtbl};
+use com_impl::{ComImpl, VTable, Refcount};
+use log::{trace, warn};
+use std::ptr::NonNull;
+use wio::com::ComPtr;
+use async_channel::Sender;
+use crate::result_enumerator::IWbemClassWrapper;
+
+/// Implementation for IWbemObjectSink.
+/// This [Sink] receives asynchronously the result of the query,
+/// through Indicate calls. When finished,the SetStatus method
+/// is called.
+/// [Sink]: https://en.wikipedia.org/wiki/Sink_(computing)
+/// # https://docs.microsoft.com/fr-fr/windows/win32/wmisdk/example--getting-wmi-data-from-the-local-computer-asynchronously
+#[repr(C)]
+#[derive(ComImpl)]
+#[interfaces(IWbemObjectSink)]
+pub struct QuerySink {
+    vtbl: VTable<IWbemObjectSinkVtbl>,
+    refcount: Refcount,
+    sender: Sender<Vec<IWbemClassWrapper>>,
+}
+
+impl QuerySink {
+    pub fn new(sender: Sender<Vec<IWbemClassWrapper>>) -> ComPtr<IWbemObjectSink> {
+        let ptr = QuerySink::create_raw(sender);
+        let ptr = ptr as *mut IWbemObjectSink;
+        unsafe { ComPtr::from_raw(ptr) }
+    }
+}
+
+#[com_impl::com_impl]
+unsafe impl IWbemObjectSink for QuerySink {
+    pub unsafe fn indicate(
+        &self,
+        lObjectCount: c_long,
+        apObjArray: *mut *mut IWbemClassObject
+    ) -> HRESULT {
+        trace!("Indicate call with {} objects", lObjectCount);
+        // apObjArray The array memory itself is read-only, and is owned by the caller of the method.
+        // Call AddRef on each element of apObjArray to borrow.
+        // https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemobjectsink-indicate
+
+        // TODO: Document when ObjectCount is <=0
+        if lObjectCount <= 0 {
+            return WBEM_NO_ERROR as i32;
+        }
+
+        let lObjectCount = lObjectCount as usize;
+        let tx = self.sender.clone();
+        let mut result = Vec::<IWbemClassWrapper>::with_capacity(lObjectCount);
```

Comment:
Any reason not to use a `Sender<IWbemClassWrapper>` directly, and send each object directly to the channel?

Reply:
I see 2 reasons by now :
> Sink implementations should process the event notification within 100 MSEC because the WMI thread that delivers the event notification cannot do other work until the sink object has completed processing. If the notification requires a large amount of processing, the sink can use an internal queue for another thread to handle the processing. 

(https://docs.microsoft.com/en-us/windows/win32/api/wbemcli/nf-wbemcli-iwbemobjectsink-indicate) => maybe It would take too much time to call the transmitter on each element, but I could be wrong.

The other reason would be that I don't see yet the need to iterate on the receiver side on each element.

It's not set in stone yet 


BTW, the way I understand the previous quote, I think the Indicate function won't be called from 2 different threads, but I have to confirm this. I don't think it will impact my implementation anyway, but that could be a good news

Reply:
I think that `send`ing to the channel should be almost immediate as it's not blocking (or at most on the order of a millisecond).

> The other reason would be that I don't see yet the need to iterate on the receiver side on each element.

If I understand correctly, indicate can be called multiple times, so if you use the vector the other side will receive an entire vector each time, and will need to iterate on it and return items from it, before moving on to the next vector. It seem simpler to have a "flat" channel so both side just send/recv a single object at a time.